src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs

```diff
@@ -0,0 +1,424 @@
+            IntPtr sslPtr = context.sslPtr;
+            IntPtr cipherPtr = libssl.SSL_get_current_cipher(sslPtr);
+            var cipher = new libssl.SSL_CIPHER();
+            if (IntPtr.Zero != cipherPtr)
+            {
+                cipher = Marshal.PtrToStructure<libssl.SSL_CIPHER>(cipherPtr);
+            }
+
+            return cipher;
+        }
+
+        internal static byte[] GetCertificateRawData(IntPtr certPtr)
+        {
+            byte[] buffer = null;
+            unsafe
+            {
+                byte* bufferPtr = null;
+                int len = libssl.i2d_X509(certPtr, ref bufferPtr);
+                if (len > 0)
+                {
+                    buffer = new byte[len];
+                    fixed (byte* pinnedBuffer = buffer)
+                    {
+                        bufferPtr = pinnedBuffer;   // TODO: creating temp var just in case
+                        len = libssl.i2d_X509(certPtr, ref bufferPtr);
+                    }
+                }
+                if (len <= 0)
+                {
+                    throw CreateSslException("Failed to get certificate raw data");
+                }
+            }
+
+            return buffer;
+        }
+
+        internal static void FreeCertificate(IntPtr certPtr)
+        {
+            if (IntPtr.Zero != certPtr)
+            {
+                libssl.X509_free(certPtr);
+            }
+        }
+
+        internal static void FreeSslContext(IntPtr sslContextPtr)
+        {
+            if (IntPtr.Zero == sslContextPtr)
+            {
+                return;
+            }
+
+            SslContext context = Marshal.PtrToStructure<SslContext>(sslContextPtr);
+            Disconnect(context.sslPtr);
+            Marshal.FreeHGlobal(sslContextPtr);
+            sslContextPtr = IntPtr.Zero;
+        }
+
+        #endregion
+
+        #region private methods
+
+        private static IntPtr GetSslMethod(bool isServer, long options)
+        {
+            long protocolMask = libssl.Options.SSL_OP_NO_SSLv2 | libssl.Options.SSL_OP_NO_SSLv3 |
+                                libssl.Options.SSL_OP_NO_TLSv1 | libssl.Options.SSL_OP_NO_TLSv1_1 |
+                                libssl.Options.SSL_OP_NO_TLSv1_2;
+            options &= protocolMask;
+            Debug.Assert(options != protocolMask, "All protocols are disabled");
+
+            var noSsl2 = (options & libssl.Options.SSL_OP_NO_SSLv2) != 0;
+            var noSsl3 = (options & libssl.Options.SSL_OP_NO_SSLv3) != 0;
+            var noTls10 = (options & libssl.Options.SSL_OP_NO_TLSv1) != 0;
+            var noTls11 = (options & libssl.Options.SSL_OP_NO_TLSv1_1) != 0;
+            var noTls12 = (options & libssl.Options.SSL_OP_NO_TLSv1_2) != 0;
+
+            var method = IntPtr.Zero;
+            if (noSsl2 && noSsl3 && noTls11 && noTls12)
+            {
+                method = isServer ? libssl.TLSv1_server_method() : libssl.TLSv1_client_method();
```

Comment:
There's also `TLSv1_method`, which leaves the option open for client or server; the call to `SSL_set_connect_state` vs `SSL_set_accept_state` later.  That would save on a bunch of conditional expressions in this method.